apps/omg_watcher_info/lib/omg_watcher_info/utxo_selection.ex

```diff
@@ -17,215 +17,163 @@ defmodule OMG.WatcherInfo.UtxoSelection do
+  - Excludes currenices not already used in the transaction and UTXOs in the selected set.
+  - Prioritises currencies that have the largest number of UTXOs
+  - Sorts by ascending order of UTXO value within the currency groupings ("dust first").
   """
-  @spec create_advice(%{Transaction.Payment.currency() => list(%DB.TxOutput{})}, order_t()) :: advice_t()
-  def create_advice(utxos, %{owner: owner, payments: payments, fee: fee} = order) do
-    needed_funds = needed_funds(payments, fee)
-    token_utxo_selection = select_utxo(utxos, needed_funds)
-
-    with {:ok, funds} <- funds_sufficient?(token_utxo_selection) do
-      utxo_count =
-        funds
-        |> Stream.map(fn {_, utxos} -> length(utxos) end)
-        |> Enum.sum()
-
-      if utxo_count <= Transaction.Payment.max_inputs(),
-        do: create_transaction(funds, order) |> respond(:complete),
-        else: create_merge(owner, funds) |> respond(:intermediate)
+  @spec prioritize_merge_utxos(utxos_map_t(), utxos_map_t()) :: utxo_list_t()
+  def prioritize_merge_utxos(utxos, selected_utxos) do
+    selected_utxos
+    |> Enum.flat_map(fn {_ccy, utxos} -> utxos end)
+    |> Enum.reduce(%{}, fn utxo, acc -> Map.put(acc, utxo.child_chain_utxohash, true) end)
+    |> case do
+      hashes_map when map_size(hashes_map) == 0 ->
+        []
+
+      hashes_map ->
+        selected_utxos
+        |> Enum.map(&prioritize_utxos_by_currency(&1, utxos, hashes_map))
+        |> Enum.sort_by(&length/1, :desc)
+        |> Enum.map(fn ccy_group -> Enum.slice(ccy_group, 0, 3) end)
+        |> List.flatten()
     end
   end
 
-  # Given available Utxo set and needed amount, we try to find an Utxo which fully satisfies the payment (without
-  # the change). If this fails, we start to collect Utxos (starting from largest amount) which will cover the payment.
-  # We return {token, {change, [utxos for payment]}}, change > 0 means insufficient funds.
-  # NOTE: order of Utxo list is implicitly assumed for the algorithm to work deterministically,
-  # see: `OMG.WatcherInfo.DB.TxOutput.get_sorted_grouped_utxos`
-  @spec select_utxo(%{Transaction.Payment.currency() => list(%DB.TxOutput{})}, %{
-          Transaction.Payment.currency() => pos_integer()
-        }) ::
-          list({Transaction.Payment.currency(), {integer, list(%DB.TxOutput{})}})
-  defp select_utxo(utxos, needed_funds) do
-    Enum.map(needed_funds, fn {token, need} ->
-      token_utxos = Map.get(utxos, token, [])
-
-      {token,
-       case Enum.find(token_utxos, fn %DB.TxOutput{amount: amount} -> amount == need end) do
-         nil ->
-           Enum.reduce_while(token_utxos, {need, []}, fn
-             _, {need, acc} when need <= 0 -> {:halt, {need, acc}}
-             %DB.TxOutput{amount: amount} = utxo, {need, acc} -> {:cont, {need - amount, [utxo | acc]}}
-           end)
-
-         utxo ->
-           {0, [utxo]}
-       end}
+  defp prioritize_utxos_by_currency({currency, _utxos}, utxos, selected_utxo_hashes) do
```

Comment:
I have now repeated this sooo many times :)))
don't mix private and public functions